src/SessionSetup/RetrieveInstances.cpp

```diff
@@ -87,70 +89,72 @@ Future<ErrorMessageOr<QVector<Instance>>> RetrieveInstancesImpl::LoadInstancesWi
+
+  const Future<ErrorMessageOr<Project>> default_project_future =
+      ggp_client_->GetDefaultProjectAsync();
+
+  const Future<ErrorMessageOr<std::pair<QVector<Instance>, std::optional<Project>>>>
+      instance_list_future = orbit_base::UnwrapFuture(
+          ggp_client_->GetInstancesAsync(scope, project)
+              .Then(main_thread_executor_,
+                    [this, scope, project](ErrorMessageOr<QVector<Instance>> instances)
+                        -> Future<
+                            ErrorMessageOr<std::pair<QVector<Instance>, std::optional<Project>>>> {
+                      if (instances.has_value()) {
+                        return {std::make_pair(instances.value(), project)};
+                      }
+                      if (project == std::nullopt ||
+                          !absl::StrContains(instances.error().message(), "it may not exist")) {
+                        return {instances.error()};
+                      }
+                      // If the error message contains "it may not exist", then the project may not
+                      // exist anymore and the call to ggp is retried without a project (default
+                      // project)
+                      return ggp_client_->GetInstancesAsync(scope, std::nullopt)
+                          .ThenIfSuccess(
+                              main_thread_executor_,
+                              [](const QVector<Instance>& instances)
+                                  -> ErrorMessageOr<
+                                      std::pair<QVector<Instance>, std::optional<Project>>> {
+                                return std::make_pair(instances, std::nullopt);
+                              });
+                    }));
+
+  Future<std::tuple<ErrorMessageOr<QVector<Project>>, ErrorMessageOr<Project>,
+                    ErrorMessageOr<std::pair<QVector<Instance>, std::optional<Project>>>>>
+      combined_future = orbit_base::JoinFutures(project_list_future, default_project_future,
+                                                instance_list_future);
 
   return combined_future.Then(
       main_thread_executor_,
-      [this](std::vector<ErrorMessageOr<void>> result_vector)
-          -> ErrorMessageOr<LoadProjectsAndInstancesResult> {
-        // Remove all that are not errors
-        result_vector.erase(
-            std::remove_if(result_vector.begin(), result_vector.end(),
-                           [](const ErrorMessageOr<void>& result) { return !result.has_error(); }),
-            result_vector.end());
-
-        if (result_vector.empty()) {
-          return projects_and_instances_load_result_;
+      [](std::tuple<ErrorMessageOr<QVector<Project>>, ErrorMessageOr<Project>,
+                    ErrorMessageOr<std::pair<QVector<Instance>, std::optional<Project>>>>
+             result_tuple) -> ErrorMessageOr<LoadProjectsAndInstancesResult> {
+        std::vector<ErrorMessage> errors;
+        if (std::get<0>(result_tuple).has_error()) {
```

Comment:
I think readability can be improved by using structured bindings. Something like:
```
const auto& [projects_or_error, default_project_or_error, instances_or_error] = result_tuple;
```

Reply:
I forgot that this exists / use this too little to have remember it. Great suggestion! I like this a lot. In case there are other similar things that are applicable, please let me know